src/block_cache/bt_io.c

```diff
@@ -338,14 +359,53 @@ __wt_bt_write(WT_SESSION_IMPL *session, WT_ITEM *buf, uint8_t *addr, size_t *add
     }
 
     /* Call the block manager to write the block. */
+    timer = WT_STAT_ENABLED(session) && !F_ISSET(session, WT_SESSION_INTERNAL);
+    time_start = timer ? __wt_clock(session) : 0;
     WT_ERR(checkpoint ? bm->checkpoint(bm, session, ip, btree->ckpt, data_checksum) :
                         bm->write(bm, session, ip, addr, addr_sizep, data_checksum, checkpoint_io));
+    if (timer) {
+        time_stop = __wt_clock(session);
+        time_diff = WT_CLOCKDIFF_US(time_stop, time_start);
+        WT_STAT_CONN_INCR(session, cache_write_app_count);
+        WT_STAT_CONN_INCRV(session, cache_write_app_time, time_diff);
+        WT_STAT_SESSION_INCRV(session, write_time, time_diff);
+    }
+
+    /*
+     * The page image must have a proper write generation number before writing it to disk. The page
+     * images that are created during recovery may have the write generation number less than the
+     * btree base write generation number, so don't verify it.
+     */
+    dsk = ip->mem;
+    WT_ASSERT(session, dsk->write_gen != 0);
 
     WT_STAT_CONN_DATA_INCR(session, cache_write);
     WT_STAT_CONN_DATA_INCRV(session, cache_bytes_write, dsk->mem_size);
     WT_STAT_SESSION_INCRV(session, bytes_write, dsk->mem_size);
     (void)__wt_atomic_add64(&S2C(session)->cache->bytes_written, dsk->mem_size);
 
+    /*
+     * Store a copy of the compressed buffer in the block cache.
+     *
+     * Optional if the write is part of a checkpoint. Hot blocks get written and over-written a lot
+     * as part of checkpoint, so we don't want to cache them, because (a) they are in the in-memory
+     * cache anyway, and (b) they are likely to be overwritten again in the next checkpoint. Writes
+     * that are not part of checkpoint I/O are done in the service of eviction. Those are the blocks
+     * that the in-memory cache would like to keep but can't, and we definitely want to keep them.
+     *
+     * Optional on normal writes (vs. reads) if the no-write-allocate setting is on.
+     *
+     * Ignore the final checkpoint writes.
+     */
+    if (blkcache->type == BLKCACHE_UNCONFIGURED)
+        ;
+    else if (!blkcache->cache_on_checkpoint && checkpoint_io)
+        WT_STAT_CONN_INCR(session, block_cache_bypass_chkpt);
+    else if (!blkcache->cache_on_writes)
+        WT_STAT_CONN_INCR(session, block_cache_bypass_writealloc);
+    else if (!checkpoint)
+        WT_ERR(__wt_blkcache_put(session, compressed ? ctmp : buf, addr, *addr_sizep, true));
```

Comment:
How paranoid should we be?  If we decide not to cache a block we could assert that there isn't a version of that block already in the cache.

Reply:
We could, but that's going to be expensive. Can you please be specific about the failure mode that worries you, that you think we should be concerned about?

In general, I'm expecting to move away from any checking at all of the cache: if we can stop storing overflow pages in the cache, there's no reason to check for blocks already being in the cache, and we can do blind inserts.

Reply:
As a general matter, I was thinking that when you treat a cache as write-around you want to ensure that a write invalidates any now-stale content in the cache.  This shouldn't be a concern in WT because we always write to free space.  But if we someday had a bug in evicting freed pages from the cache? 

I realize on spelling this out, however, that checksums are sufficient to detect this type of issue.

Reply:
And, in fact, checksums will detect pretty much all the ways the cache might mangle or misplace blocks.  After all it's just another storage device that we can't trust :-)